File: LeetCode/354.Russian_Doll_Envelopes/354.Russian_Doll_Envelopes.java
class Solution {
    public int maxEnvelopes(int[][] envelopes) {
        Arrays.sort(envelopes, (a, b) -> {
            return a[0] == b[0] ? b[1] - a[1] : a[0] - b[0];
        });

        int n = envelopes.length;
        int[] tops = new int[n];
        int piles = 0;
        for (int[] e : envelopes) {
            int num = e[1];

            int left = 0, right = piles;
            while (left < right) {
                int mid = (left + right) / 2;
                if (tops[mid] < num) {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }

            if (left == piles)
                piles += 1;
            tops[left] = num;
        }

        return piles;
    }
}

File: LeetCode/354.Russian_Doll_Envelopes/354.Russian_Doll_Envelopes.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import bisect
from typing import List


class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        envelopes.sort(key=lambda e: (e[0], -e[1]))

        heights = [e[1] for e in envelopes]

        tails = []
        for h in heights:
            idx = bisect.bisect_left(tails, h)
            if idx == len(tails):
                tails.append(h)
            else:
                tails[idx] = h

        return len(tails)
